hack/generator/pkg/codegen/pipeline_create_arm_types.go

```diff
@@ -0,0 +1,464 @@
+			if err != nil {
+				return astmodel.TypeName{}, nil, err
+			}
+
+			armDef, ok := kubeNameToArmDefs[kubernetesSpecDef.Name()]
+			if !ok {
+				return astmodel.TypeName{}, nil, errors.Errorf("couldn't find arm def matching kube def %q", kubernetesSpecDef.Name())
+			}
+
+			result, err := addArmConversionInterface(*kubernetesSpecDef, armDef, idFactory, true)
+			if err != nil {
+				return astmodel.TypeName{}, nil, err
+			}
+
+			return result.Name(), result, nil
+		},
+		// Other defs handler
+		func(def astmodel.TypeDefinition) (*astmodel.TypeDefinition, error) {
+			armDef, ok := kubeNameToArmDefs[def.Name()]
+			if !ok {
+				return nil, errors.Errorf("couldn't find arm def matching kube def %q", def.Name())
+			}
+
+			return addArmConversionInterface(
+				def,
+				armDef,
+				idFactory,
+				false)
+		})
+}
+
+func iterDefs(
+	definitions astmodel.Types,
+	resourceHandler func(name astmodel.TypeName, resourceType *astmodel.ResourceType) (astmodel.TypeName, *astmodel.TypeDefinition, error),
+	otherDefsHandler func(def astmodel.TypeDefinition) (*astmodel.TypeDefinition, error)) (astmodel.Types, error) {
+
+	newDefs := make(astmodel.Types)
+	actionedDefs := make(map[astmodel.TypeName]struct{})
+
+	// Do all the resources first - this ensures we avoid handling a spec before we've processed
+	// its associated resource.
+	for _, def := range definitions {
+		// Special handling for resources because we need to modify their specs with extra properties
+		if resourceType, ok := def.Type().(*astmodel.ResourceType); ok {
+			specTypeName, newDef, err := resourceHandler(def.Name(), resourceType)
+			if err != nil {
+				return nil, err
+			}
+
+			newDefs.Add(*newDef)
+			actionedDefs[specTypeName] = struct{}{}
+		}
+	}
+
+	// Process the remaining definitions
+	for _, def := range definitions {
+		// If it's a type which has already been handled (specs from above), skip it
+		if _, ok := actionedDefs[def.Name()]; ok {
+			continue
+		}
+
+		// Note: We would need to do something about type aliases here if they weren't already
+		// removed earlier in the pipeline
+
+		// Other types can be reused in both the Kube type graph and the ARM type graph, for
+		// example enums which are effectively primitive types, all primitive types, etc.
+		_, ok := def.Type().(*astmodel.ObjectType)
+		if !ok {
+			continue
+		}
+
+		newDef, err := otherDefsHandler(def)
+		if err != nil {
+			return nil, err
+		}
+		newDefs.Add(*newDef)
+	}
+
+	return newDefs, nil
+}
+
+func removeValidations(t *astmodel.ObjectType) (*astmodel.ObjectType, error) {
+	result := t
+
+	for _, p := range t.Properties() {
+		result = result.WithoutProperty(p.PropertyName())
+	}
+
+	for _, p := range t.Properties() {
+		p = p.WithoutValidation()
+		result = result.WithProperty(p)
+	}
+
+	return result, nil
+}
+
+func createArmTypeName(name astmodel.TypeName) astmodel.TypeName {
+	return astmodel.MakeTypeName(name.PackageReference, name.Name()+"Arm")
```

Comment:
If this happened, the type which was returned would cause a panic when we attempted to perform `result := astmodel.TypesSetDisjointUnion(armTypes, kubeTypes)` (since there would be a collision), so it's definitely not happening now and we would certainly know if it did.

Solving that problem in a way that is "good" might be a bit tricky... for now I've made a note that we should think about what we might want to do in that case, but I think it makes sense to punt on actually worrying about it right now since it doesn't happen and I'd imagine an actual solution is something like a:
`TypeNamer` or `IdFactory` that has context (by package?) so that it can detect collisions and push us towards better names.

Reply:
I think that makes sense - and as long as it doesn't affect the name of the k8s types I don't think we'll find that we're painted into a corner, since nothing other than the code knows the names of the arm types we're free to fix that however seems best if/when it comes up.

Reply:
Resolving this as it seems there's agreement - we'll cross this bridge if/when we come to it.